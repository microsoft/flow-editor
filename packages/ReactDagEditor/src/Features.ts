export enum GraphFeatures {
  nodeDraggable = "nodeDraggable",
  nodeResizable = "nodeResizable",
  clickNodeToSelect = "clickNodeToSelect",
  panCanvas = "panCanvas",
  multipleSelect = "multipleSelect",
  lassoSelect = "lassoSelect",
  delete = "delete",
  addNewNodes = "addNewNodes",
  addNewEdges = "addNewEdges",
  addNewPorts = "addNewPorts",
  autoFit = "autoFit",
  canvasHorizontalScrollable = "canvasHorizontalScrollable",
  canvasVerticalScrollable = "canvasVerticalScrollable",
  nodeHoverView = "nodeHoverView",
  portHoverView = "portHoverView",
  addEdgesByKeyboard = "addEdgesByKeyboard",
  a11yFeatures = "a11yFeatures",
  editNode = "editNode",
  autoAlign = "autoAlign",
  undoStack = "undoStack",
  ctrlKeyZoom = "ctrlKeyZoom",
  limitBoundary = "limitBoundary",
  editEdge = "editEdge"
}

export const allFeatures = new Set([
  GraphFeatures.nodeDraggable,
  GraphFeatures.nodeResizable,
  GraphFeatures.clickNodeToSelect,
  GraphFeatures.panCanvas,
  GraphFeatures.multipleSelect,
  GraphFeatures.lassoSelect,
  GraphFeatures.delete,
  GraphFeatures.addNewNodes,
  GraphFeatures.addNewEdges,
  GraphFeatures.addNewPorts,
  GraphFeatures.canvasHorizontalScrollable,
  GraphFeatures.canvasVerticalScrollable,
  GraphFeatures.nodeHoverView,
  GraphFeatures.portHoverView,
  GraphFeatures.addEdgesByKeyboard,
  GraphFeatures.a11yFeatures,
  GraphFeatures.autoFit,
  GraphFeatures.editNode,
  GraphFeatures.autoAlign,
  GraphFeatures.undoStack,
  GraphFeatures.ctrlKeyZoom,
  GraphFeatures.limitBoundary,
  GraphFeatures.editEdge
]);

export const defaultFeatures = new Set([
  GraphFeatures.nodeDraggable,
  GraphFeatures.nodeResizable,
  GraphFeatures.clickNodeToSelect,
  GraphFeatures.panCanvas,
  GraphFeatures.multipleSelect,
  GraphFeatures.delete,
  GraphFeatures.addNewNodes,
  GraphFeatures.addNewEdges,
  GraphFeatures.addNewPorts,
  GraphFeatures.canvasHorizontalScrollable,
  GraphFeatures.canvasVerticalScrollable,
  GraphFeatures.nodeHoverView,
  GraphFeatures.portHoverView,
  GraphFeatures.addEdgesByKeyboard,
  GraphFeatures.a11yFeatures,
  GraphFeatures.editNode,
  GraphFeatures.autoAlign,
  GraphFeatures.undoStack,
  GraphFeatures.ctrlKeyZoom,
  GraphFeatures.limitBoundary
]);

export const dataReadonlyMode = new Set([
  GraphFeatures.nodeDraggable,
  GraphFeatures.nodeResizable,
  GraphFeatures.clickNodeToSelect,
  GraphFeatures.panCanvas,
  GraphFeatures.multipleSelect,
  GraphFeatures.canvasHorizontalScrollable,
  GraphFeatures.canvasVerticalScrollable,
  GraphFeatures.nodeHoverView,
  GraphFeatures.portHoverView,
  GraphFeatures.a11yFeatures,
  GraphFeatures.ctrlKeyZoom,
  GraphFeatures.limitBoundary
]);

export const GanttChartFeatures = new Set([
  GraphFeatures.clickNodeToSelect,
  GraphFeatures.canvasHorizontalScrollable,
  GraphFeatures.canvasVerticalScrollable,
  GraphFeatures.nodeHoverView,
  GraphFeatures.portHoverView,
  GraphFeatures.a11yFeatures,
  GraphFeatures.lassoSelect,
  GraphFeatures.limitBoundary
]);

export const previewMode = new Set([GraphFeatures.nodeHoverView, GraphFeatures.portHoverView, GraphFeatures.autoFit]);
