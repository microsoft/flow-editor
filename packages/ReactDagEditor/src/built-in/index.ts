export * from "./rect";
export * from "./line";
export * from "./defaultPort";
export * from "./InMemoryClipboard";
export * from "./InMemoryStorage";
export * from "./applyDefaultPortsPosition";
export * from "./applyHorizontalGraphPortsPosition";
