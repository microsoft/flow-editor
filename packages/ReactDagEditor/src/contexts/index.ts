export * from "./GraphConfigContext";
export * from "./ContextMenuConfigContext";
export * from "./ThemeContext";
export * from "./ViewportContext";
export * from "./GraphStateContext";
