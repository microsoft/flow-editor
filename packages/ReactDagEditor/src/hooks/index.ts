export * from "./useContextMenuConfigContext";
export * from "./useWheelHandler";
export * from "./useClientRect";
export * from "./useTheme";
export * from "./useContainerRect";
export * from "./useThrottle";
export * from "./useRenderedArea";
export * from "./useMinimapRect";
export * from "./useGraphTouchHandler";
export * from "./useSafariScale";
export * from "./useSelectBox";
export * from "./useGraphState";
export * from "./useDeferredValue";
