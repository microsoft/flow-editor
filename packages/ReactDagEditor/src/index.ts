export * from "./components";
export * from "./contexts";
export * from "./built-in";
export * from "./utils";
export * from "./hooks";
export * from "./testAPI";
export * from "./Features";
export * from "./models/EdgeModel";
export * from "./models/GraphModel";
export * from "./models/NodeModel";
export * from "./reducers/composeReducers";
export * from "./models/element-state";
export * from "./models/port";
export * from "./models/node";
export * from "./models/edge";
export * from "./models/canvas";
export * from "./models/geometry";
export * from "./models/state";
export * from "./models/event";
