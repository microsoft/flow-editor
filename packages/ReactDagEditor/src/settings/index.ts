export * from "./GraphConfig";
export * from "./IGraphClipboard";
export * from "./IGraphConfig";
