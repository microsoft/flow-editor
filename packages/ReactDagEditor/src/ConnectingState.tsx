import * as React from "react";
import { ConnectingStateContext, IConnectingStateContext } from "./contexts/ConnectingStateContext";
import { GraphModel } from "./models/GraphModel";
import { NodeModel } from "./models/NodeModel";
import { ICanvasPort } from "./models/port";
import { IConnectingState } from "./models/state";

interface IProps {
  data: GraphModel;
  connectState: IConnectingState | undefined;
}

export const ConnectingState: React.FunctionComponent<IProps> = ({ children, data, connectState }) => {
  let sourceNode: NodeModel | undefined;
  let sourcePort: ICanvasPort | undefined;
  let targetNode: NodeModel | undefined;
  let targetPort: ICanvasPort | undefined;
  if (connectState) {
    sourceNode = data.nodes.get(connectState.sourceNode);
    sourcePort = sourceNode?.getPort(connectState.sourcePort);
    targetNode = connectState.targetNode ? data.nodes.get(connectState.targetNode) : undefined;
    targetPort = connectState.targetPort ? targetNode?.getPort(connectState.targetPort) : undefined;
  }
  const context: IConnectingStateContext = React.useMemo(
    () => ({
      sourceNode,
      sourcePort,
      targetNode,
      targetPort
    }),
    [sourceNode, sourcePort, targetNode, targetPort]
  );
  return <ConnectingStateContext.Provider value={context}>{children}</ConnectingStateContext.Provider>;
};

ConnectingState.displayName = "ConnectingState";
