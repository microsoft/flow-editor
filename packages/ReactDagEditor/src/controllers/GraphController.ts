import { IDispatch } from "../contexts";
import { GraphFeatures } from "../Features";
import { IPoint } from "../models/geometry";
import { GraphModel } from "../models/GraphModel";
import { NodeModel } from "../models/NodeModel";
import { GraphBehavior, IGraphState } from "../models/state";
import { EventChannel } from "../utils/eventChannel";

/**
 * event handlers must get/set GraphBehavior immediately
 * to determine how to treat event objects (eg. preventDefault)
 * but dispatch/setState can be asynchronous
 * thus GraphBehavior is mirrored here
 */
export class GraphController {
  /**
   * since we don't have a detailed specification for touch handling
   * temporarily store the pointerId here for most single point events
   */
  public pointerId: number | null = null;
  /**
   * simulate canvas click event once
   */
  public canvasClickOnce = false;
  public nodeClickOnce: NodeModel | null = null;
  public readonly eventChannel = new EventChannel();
  public state: IGraphState;
  public UNSAFE_latestState: IGraphState;
  public dispatchDelegate: IDispatch;
  public getGlobalEventTargetImpl?: () => Window | Element | null | undefined;
  private mouseClientPoint?: IPoint;
  private behavior = GraphBehavior.default;

  public constructor(state: IGraphState, dispatch: IDispatch) {
    this.state = state;
    this.UNSAFE_latestState = state;
    this.dispatch = dispatch;
  }

  public readonly dispatch: IDispatch = (action, callback) => {
    this.dispatchDelegate(action, callback);
  };

  public setMouseClientPosition(pos: IPoint): void {
    this.mouseClientPoint = pos;
  }

  public unsetMouseClientPosition(): void {
    this.mouseClientPoint = undefined;
  }

  public getMouseClientPosition(): IPoint | undefined {
    return this.mouseClientPoint;
  }

  public getEnabledFeatures(): ReadonlySet<GraphFeatures> {
    return this.state.settings.features;
  }

  public getBehavior(): GraphBehavior {
    return this.behavior;
  }

  public setBehavior(value: GraphBehavior): void {
    this.behavior = value;
  }

  public getData(): GraphModel {
    return this.state.data.present;
  }

  public getGlobalEventTarget(): Element | Window {
    return this.getGlobalEventTargetImpl?.() ?? window;
  }
}
