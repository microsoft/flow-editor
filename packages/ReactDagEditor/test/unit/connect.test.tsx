/* eslint-disable @typescript-eslint/no-explicit-any,@typescript-eslint/no-non-null-assertion,compat/compat */
import { cleanup, render, RenderResult } from "@testing-library/react";
import * as React from "react";
import { act } from "react-dom/test-utils";
import {
  applyDefaultPortsPosition,
  Bitset,
  GraphCanvasEvent,
  GraphEdgeEvent,
  GraphModel,
  GraphNodeEvent,
  GraphNodeStatus,
  GraphPortEvent,
  GraphPortStatus,
  ICanvasData,
  ICanvasNode,
  ICanvasPort,
  IConnectingState,
  IGraphProps,
  IGraphState,
  NodeModel,
} from "../../src";
import { GraphController } from "../../src/controllers/GraphController";
import { getNearestConnectablePort } from "../../src/utils";
import { EventChannel } from "../../src/utils/eventChannel";
import { identical } from "../../src/utils/identical";
import { GraphControllerRef, TestComponent } from "../TestComponent";
import { getGraphConfig, mockClientRect, patchPointerEvent } from "../utils";

describe("test getNearestConnectablePort", () => {
  it("should return nothing with no ports", () => {
    const graphConfig = getGraphConfig();
    const node: ICanvasNode = {
      id: "0",
      width: 200,
      height: 100,
      x: 100,
      y: 100,
    };
    const data = GraphModel.fromJSON({
      nodes: [node],
      edges: [],
    });
    const port = getNearestConnectablePort({
      parentNode: NodeModel.fromJSON(node, undefined, undefined),
      clientX: 150,
      clientY: 150,
      graphConfig,
      data,
      viewport: {
        rect: mockClientRect,
        transformMatrix: [1, 0, 0, 1, 0, 0],
      },
      anotherNode: NodeModel.fromJSON(node, undefined, undefined),
      anotherPort: {
        id: "0",
        name: "",
        position: [0, 0],
        isInputDisabled: false,
        data: {
          isConnectable: false,
        },
      },
    });
    expect(port).toBeUndefined();
  });

  it("should return connectable port", () => {
    const graphConfig = getGraphConfig();
    const ports: ICanvasPort[] = applyDefaultPortsPosition([
      {
        id: "0",
        name: "",
        isInputDisabled: false,
        data: {
          isConnectable: false,
        },
      },
      {
        id: "1",
        name: "",
        isInputDisabled: false,
        data: {
          isConnectable: true,
        },
      },
      {
        id: "2",
        name: "",
        isInputDisabled: false,
        data: {
          isConnectable: true,
        },
      },
    ]);
    const node: ICanvasNode = {
      id: "0",
      width: 200,
      height: 100,
      x: 100,
      y: 100,
      ports,
    };
    const data = GraphModel.fromJSON({
      nodes: [node],
      edges: [],
    });
    const port = getNearestConnectablePort({
      parentNode: NodeModel.fromJSON(node, undefined, undefined),
      clientX: 250,
      clientY: 150,
      graphConfig,
      data,
      viewport: {
        rect: mockClientRect,
        transformMatrix: [1, 0, 0, 1, 0, 0],
      },
      anotherNode: NodeModel.fromJSON(node, undefined, undefined),
      anotherPort: {
        id: "0",
        name: "",
        position: [0, 0],
        isInputDisabled: false,
        data: {
          isConnectable: true,
        },
      },
    });
    expect(port?.id).toBe("1");
  });
});

describe("test Connecting", () => {
  beforeAll(() => {
    patchPointerEvent();
  });

  let container: RenderResult;
  let edgeWillAdd: Required<IGraphProps>["edgeWillAdd"];
  let graphController: GraphController;
  let eventChannel: EventChannel;
  let mockData: GraphModel;
  let ports: ICanvasPort[];
  let connecting: Connecting;

  const getData = () => {
    return graphController.state.data.present;
  };

  class Connecting {
    public start(node: NodeModel, port: ICanvasPort): void {
      act(() => {
        eventChannel.trigger({
          type: GraphEdgeEvent.ConnectStart,
          nodeId: node.id,
          portId: port.id,
          rawEvent: new KeyboardEvent("keydown") as any,
          clientPoint: {
            x: 190 + mockClientRect.left,
            y: 151 + mockClientRect.top,
          },
        });
      });
    }

    public attach(node: NodeModel, port: ICanvasPort): void {
      graphController.pointerId = 0;
      act(() => {
        eventChannel.trigger({
          type: GraphPortEvent.PointerEnter,
          node,
          port,
          rawEvent: new PointerEvent("pointerenter", { pointerId: 0 }),
        });
      });
    }

    public clearAttach(node: NodeModel, port: ICanvasPort): void {
      graphController.pointerId = 0;
      act(() => {
        eventChannel.trigger({
          type: GraphPortEvent.PointerLeave,
          node,
          port,
          rawEvent: new PointerEvent("pointerenter", { pointerId: 0 }),
        });
      });
    }

    public finish(): void {
      act(() => {
        eventChannel.trigger({
          type: GraphEdgeEvent.ConnectEnd,
          edgeWillAdd,
          isCancel: false,
          rawEvent: new KeyboardEvent("keydown"),
        });
      });
    }

    public get state(): IGraphState | undefined {
      return graphController.state;
    }

    public get connectState(): IConnectingState | undefined {
      return graphController.state.connectState;
    }

    public get data(): GraphModel | undefined {
      return this.state?.data.present;
    }

    public get sourceNode(): string | undefined {
      return this.connectState?.sourceNode;
    }

    public get sourcePort(): string | undefined {
      return this.connectState?.sourcePort;
    }

    public get targetNode(): string | undefined {
      return this.connectState?.targetNode;
    }

    public get targetPort(): string | undefined {
      return this.connectState?.targetPort;
    }
  }

  beforeEach(() => {
    ports = applyDefaultPortsPosition([
      {
        id: "0",
        name: "0",
        status: GraphPortStatus.Default,
        isInputDisabled: false,
        isOutputDisabled: true,
        data: {
          isConnectable: false,
        },
      },
      {
        id: "1",
        name: "1",
        status: GraphPortStatus.Default,
        isInputDisabled: true,
        isOutputDisabled: false,
        data: {
          isConnectable: true,
        },
      },
    ]);

    mockData = GraphModel.fromJSON({
      edges: [],
      nodes: [
        {
          id: "0",
          x: 50,
          y: 50,
          width: 100,
          height: 100,
          ports,
        },
        {
          id: "1",
          x: 150,
          y: 150,
          width: 100,
          height: 100,
          ports,
        },
        {
          id: "2",
          x: 250,
          y: 250,
          width: 150,
          height: 150,
          ports,
        },
      ],
    });

    edgeWillAdd = jest.fn(identical);
    const graphControllerRef = React.createRef<GraphController>();
    container = render(
      <TestComponent settings={{ graphConfig: getGraphConfig() }} data={mockData} graphProps={{ edgeWillAdd }}>
        <GraphControllerRef ref={graphControllerRef} />
      </TestComponent>
    );
    graphController = graphControllerRef.current!;
    expect(graphController).toBeDefined();
    eventChannel = graphController.eventChannel;
    expect(eventChannel).toBeDefined();
    connecting = new Connecting();
    act(() => {
      eventChannel.trigger({
        type: GraphCanvasEvent.ViewportResize,
        viewportRect: mockClientRect,
      });
    });
  });

  afterEach(cleanup);

  const getConnectLineElement = () => container.baseElement.querySelector("#markerArrow")?.parentElement?.parentElement;

  it("should render invisible lines", () => {
    expect(getConnectLineElement()).toMatchSnapshot();
  });

  it("should not connect", () => {
    act(() => {
      connecting.start(mockData.nodes.get(mockData.head!)!, ports[0]);
    });
    const port = getData().nodes.get(mockData.head!)?.ports?.[0];
    expect(Bitset.has(GraphPortStatus.Connecting)(port?.status)).toBe(true);
    expect(connecting.sourceNode).toBe("0");
    expect(connecting.sourcePort).toBe("0");
    act(() => {
      connecting.finish();
    });
    expect(getData().toJSON()).toEqual(mockData.toJSON());
    expect(connecting.sourceNode).toBeUndefined();
    expect(connecting.sourcePort).toBeUndefined();
    expect(edgeWillAdd).not.toBeCalled();
  });

  it("should attach to port", () => {
    act(() => {
      const node = mockData.nodes.get(mockData.head!)!;
      connecting.start(node, node.ports![1]);
    });
    expect(getConnectLineElement()).toMatchSnapshot();
    act(() => {
      connecting.attach(mockData.nodes.get("1")!, mockData.nodes.get("1")!.ports![1]);
    });
    expect(Bitset.has(GraphPortStatus.ConnectingAsTarget)(connecting.data?.nodes.get("1")!.ports![1].status)).toBe(
      true
    );
    expect(getConnectLineElement()).toMatchSnapshot();
    act(() => {
      connecting.clearAttach(mockData.nodes.get("1")!, mockData.nodes.get("1")!.ports![1]);
    });
    expect(getData().toJSON()).toMatchSnapshot();
    expect(getConnectLineElement()).toMatchSnapshot();
  });

  it("should add edge", () => {
    const firstNode = mockData.nodes.get(mockData.head!)!;
    act(() => {
      connecting.start(firstNode, firstNode.ports![1]);
    });
    act(() => {
      const node = mockData.nodes.get(firstNode.next!)!;
      connecting.attach(node, node.ports![1]);
    });
    act(() => {
      connecting.finish();
    });
    expect(connecting.data?.toJSON()).toMatchSnapshot();
    expect(connecting.sourceNode).toBeUndefined();
    expect(connecting.sourcePort).toBeUndefined();
    expect(connecting.targetNode).toBeUndefined();
    expect(connecting.targetPort).toBeUndefined();
    expect(edgeWillAdd).toBeCalledTimes(1);
    expect(getConnectLineElement()).toMatchSnapshot();
  });

  it("should attach nearest port by mouse, should not add edge after mouse leave", () => {
    const data: ICanvasData = {
      edges: [],
      nodes: [
        {
          id: "0",
          x: 50,
          y: 50,
          width: 100,
          height: 100,
          status: GraphNodeStatus.Default,
          ports: applyDefaultPortsPosition([
            {
              id: "0",
              name: "0",
              status: GraphPortStatus.Default,
              isInputDisabled: false,
              isOutputDisabled: true,
            },
            {
              id: "1",
              name: "1",
              status: GraphPortStatus.Default,
              isInputDisabled: true,
              isOutputDisabled: false,
            },
          ]),
        },
        {
          id: "1",
          x: 150,
          y: 150,
          width: 100,
          height: 100,
          status: GraphNodeStatus.Default,
          ports: applyDefaultPortsPosition([
            {
              id: "0",
              name: "0",
              status: GraphPortStatus.Default,
              isInputDisabled: false,
              isOutputDisabled: true,
              data: {
                isConnectable: true,
              },
            },
            {
              id: "1",
              name: "1",
              status: GraphPortStatus.Default,
              isInputDisabled: false,
              isOutputDisabled: true,
              data: {
                isConnectable: true,
              },
            },
            {
              id: "2",
              name: "2",
              status: GraphPortStatus.Default,
              isInputDisabled: false,
              isOutputDisabled: true,
              data: {
                isConnectable: true,
              },
            },
            {
              id: "3",
              name: "3",
              status: GraphPortStatus.Default,
              isInputDisabled: true,
              isOutputDisabled: false,
              data: {
                isConnectable: true,
              },
            },
          ]),
        },
      ],
    };
    act(() => {
      graphController.dispatch({
        type: GraphCanvasEvent.SetData,
        data: GraphModel.fromJSON(data),
      });
    });
    act(() => {
      connecting.start(NodeModel.fromJSON(data.nodes[0], undefined, "1"), data.nodes[0].ports![1]);
    });
    expect(connecting.targetNode).toBeUndefined();
    expect(connecting.targetPort).toBeUndefined();

    act(() => {
      eventChannel.trigger({
        type: GraphNodeEvent.PointerEnter,
        node: NodeModel.fromJSON(data.nodes[1], "0", undefined),
        rawEvent: new PointerEvent("pointerenter", {
          pointerId: 0,
          clientX: 210 + mockClientRect.left,
          clientY: 160 + mockClientRect.top,
        }),
      });
    });
    expect(connecting.targetNode).toBe("1");
    expect(connecting.targetPort).toBe("0");

    act(() => {
      eventChannel.trigger({
        type: GraphNodeEvent.PointerLeave,
        node: NodeModel.fromJSON(data.nodes[1], "0", undefined),
        rawEvent: new PointerEvent("pointerenter", {
          pointerId: 0,
          clientX: 210 + mockClientRect.left,
          clientY: 160 + mockClientRect.top,
        }),
      });
    });
    expect(connecting.targetNode).toBeUndefined();
    expect(connecting.targetPort).toBeUndefined();

    act(() => {
      connecting.finish();
    });
    expect(connecting.data?.toJSON()).toEqual(data);
  });

  it("should attach to port when mouse enter a connectable port, should not add edge after mouse leave", () => {
    act(() => {
      connecting.start(mockData.nodes.get("0")!, mockData.nodes.get("0")!.ports![1]);
    });
    expect(connecting.targetNode).toBeUndefined();
    expect(connecting.targetPort).toBeUndefined();
    expect(getConnectLineElement()).toMatchSnapshot();

    act(() => {
      connecting.attach(mockData.nodes.get("1")!, mockData.nodes.get("1")!.ports![0]);
    });
    expect(connecting.targetNode).toBeUndefined();
    expect(connecting.targetPort).toBeUndefined();
    expect(getConnectLineElement()).toMatchSnapshot();

    act(() => {
      const node = mockData.nodes.get("1")!;
      const port = mockData.nodes.get("1")!.ports![0];
      eventChannel.trigger({
        type: GraphPortEvent.PointerLeave,
        rawEvent: new PointerEvent("pointerleave"),
        node,
        port,
      });
    });
    expect(getConnectLineElement()).toMatchSnapshot();

    act(() => {
      connecting.attach(mockData.nodes.get("1")!, mockData.nodes.get("1")!.ports![1]);
    });
    expect(connecting.targetNode).toBe("1");
    expect(connecting.targetPort).toBe("1");
    expect(getConnectLineElement()).toMatchSnapshot();

    act(() => {
      const node = mockData.nodes.get("1")!;
      const port = mockData.nodes.get("1")!.ports![1];
      eventChannel.trigger({
        type: GraphPortEvent.PointerLeave,
        rawEvent: new PointerEvent("pointerleave"),
        node,
        port,
      });
    });
    expect(connecting.targetNode).toBeUndefined();
    expect(connecting.targetPort).toBeUndefined();
    expect(getConnectLineElement()).toMatchSnapshot();

    act(() => {
      connecting.finish();
    });
    expect(connecting.data?.toJSON()).toEqual(mockData.toJSON());
  });
});
