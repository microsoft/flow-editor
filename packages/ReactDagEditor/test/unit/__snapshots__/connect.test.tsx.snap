// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Connecting should add edge 1`] = `undefined`;

exports[`test Connecting should add edge: events 1`] = `
Array [
  "[Canvas]ViewportResize",
  "[Canvas]VirtualizationRecalculated",
  "[Canvas]ViewportResize",
  "[Edge]ConnectStart",
  "[Port]PointerEnter",
  "[Edge]ConnectEnd",
]
`;

exports[`test Connecting should attach nearest port by mouse, should not add edge after mouse leave: events 1`] = `
Array [
  "[Canvas]ViewportResize",
  "[Canvas]VirtualizationRecalculated",
  "[Canvas]ViewportResize",
  "[Edge]ConnectStart",
  "[Node]PointerEnter",
  "[Node]PointerLeave",
  "[Edge]ConnectEnd",
]
`;

exports[`test Connecting should attach to port 1`] = `
<g>
  <defs>
    <marker
      id="markerArrow"
      markerHeight="10"
      markerUnits="strokeWidth"
      markerWidth="10"
      orient="auto"
      refX="6"
      refY="5"
    >
      <path
        d="M0,0 L6,5 L0,10"
        style="stroke: #0078D4; fill: none;"
      />
    </marker>
  </defs>
  <line
    marker-end="url(#markerArrow)"
    style="stroke: #0078D4; fill: none; stroke-dasharray: 4,4;"
    x1="190"
    x2="190"
    y1="150"
    y2="151"
  />
  <path
    d="M190,150C190,135,190,170,190,155"
    style="stroke: none; fill: none;"
  />
</g>
`;

exports[`test Connecting should attach to port 2`] = `
<g>
  <defs>
    <marker
      id="markerArrow"
      markerHeight="10"
      markerUnits="strokeWidth"
      markerWidth="10"
      orient="auto"
      refX="6"
      refY="5"
    >
      <path
        d="M0,0 L6,5 L0,10"
        style="stroke: none; fill: none;"
      />
    </marker>
  </defs>
  <line
    marker-end="url(#markerArrow)"
    style="stroke: none; fill: none; stroke-dasharray: 4,4;"
    x1="190"
    x2="190"
    y1="150"
    y2="151"
  />
  <path
    d="M290,250C290,202.5,190,202.5,190,155"
    style="stroke: #0078D4; fill: none;"
  />
</g>
`;

exports[`test Connecting should attach to port 3`] = `
<g>
  <defs>
    <marker
      id="markerArrow"
      markerHeight="10"
      markerUnits="strokeWidth"
      markerWidth="10"
      orient="auto"
      refX="6"
      refY="5"
    >
      <path
        d="M0,0 L6,5 L0,10"
        style="stroke: #0078D4; fill: none;"
      />
    </marker>
  </defs>
  <line
    marker-end="url(#markerArrow)"
    style="stroke: #0078D4; fill: none; stroke-dasharray: 4,4;"
    x1="190"
    x2="190"
    y1="150"
    y2="151"
  />
  <path
    d="M190,150C190,135,190,170,190,155"
    style="stroke: none; fill: none;"
  />
</g>
`;

exports[`test Connecting should attach to port when mouse enter a connectable port, should not add edge after mouse leave 1`] = `
<g>
  <defs>
    <marker
      id="markerArrow"
      markerHeight="10"
      markerUnits="strokeWidth"
      markerWidth="10"
      orient="auto"
      refX="6"
      refY="5"
    >
      <path
        d="M0,0 L6,5 L0,10"
        style="stroke: #0078D4; fill: none;"
      />
    </marker>
  </defs>
  <line
    marker-end="url(#markerArrow)"
    style="stroke: #0078D4; fill: none; stroke-dasharray: 4,4;"
    x1="190"
    x2="190"
    y1="150"
    y2="151"
  />
  <path
    d="M190,150C190,135,190,170,190,155"
    style="stroke: none; fill: none;"
  />
</g>
`;

exports[`test Connecting should attach to port when mouse enter a connectable port, should not add edge after mouse leave 2`] = `
<g>
  <defs>
    <marker
      id="markerArrow"
      markerHeight="10"
      markerUnits="strokeWidth"
      markerWidth="10"
      orient="auto"
      refX="6"
      refY="5"
    >
      <path
        d="M0,0 L6,5 L0,10"
        style="stroke: #0078D4; fill: none;"
      />
    </marker>
  </defs>
  <line
    marker-end="url(#markerArrow)"
    style="stroke: #0078D4; fill: none; stroke-dasharray: 4,4;"
    x1="190"
    x2="190"
    y1="150"
    y2="151"
  />
  <path
    d="M190,150C190,135,190,170,190,155"
    style="stroke: none; fill: none;"
  />
</g>
`;

exports[`test Connecting should attach to port when mouse enter a connectable port, should not add edge after mouse leave 3`] = `
<g>
  <defs>
    <marker
      id="markerArrow"
      markerHeight="10"
      markerUnits="strokeWidth"
      markerWidth="10"
      orient="auto"
      refX="6"
      refY="5"
    >
      <path
        d="M0,0 L6,5 L0,10"
        style="stroke: #0078D4; fill: none;"
      />
    </marker>
  </defs>
  <line
    marker-end="url(#markerArrow)"
    style="stroke: #0078D4; fill: none; stroke-dasharray: 4,4;"
    x1="190"
    x2="190"
    y1="150"
    y2="151"
  />
  <path
    d="M190,150C190,135,190,170,190,155"
    style="stroke: none; fill: none;"
  />
</g>
`;

exports[`test Connecting should attach to port when mouse enter a connectable port, should not add edge after mouse leave 4`] = `
<g>
  <defs>
    <marker
      id="markerArrow"
      markerHeight="10"
      markerUnits="strokeWidth"
      markerWidth="10"
      orient="auto"
      refX="6"
      refY="5"
    >
      <path
        d="M0,0 L6,5 L0,10"
        style="stroke: none; fill: none;"
      />
    </marker>
  </defs>
  <line
    marker-end="url(#markerArrow)"
    style="stroke: none; fill: none; stroke-dasharray: 4,4;"
    x1="190"
    x2="190"
    y1="150"
    y2="151"
  />
  <path
    d="M290,250C290,202.5,190,202.5,190,155"
    style="stroke: #0078D4; fill: none;"
  />
</g>
`;

exports[`test Connecting should attach to port when mouse enter a connectable port, should not add edge after mouse leave 5`] = `
<g>
  <defs>
    <marker
      id="markerArrow"
      markerHeight="10"
      markerUnits="strokeWidth"
      markerWidth="10"
      orient="auto"
      refX="6"
      refY="5"
    >
      <path
        d="M0,0 L6,5 L0,10"
        style="stroke: #0078D4; fill: none;"
      />
    </marker>
  </defs>
  <line
    marker-end="url(#markerArrow)"
    style="stroke: #0078D4; fill: none; stroke-dasharray: 4,4;"
    x1="190"
    x2="190"
    y1="150"
    y2="151"
  />
  <path
    d="M190,150C190,135,190,170,190,155"
    style="stroke: none; fill: none;"
  />
</g>
`;

exports[`test Connecting should attach to port when mouse enter a connectable port, should not add edge after mouse leave: events 1`] = `
Array [
  "[Canvas]ViewportResize",
  "[Canvas]VirtualizationRecalculated",
  "[Canvas]ViewportResize",
  "[Edge]ConnectStart",
  "[Port]PointerEnter",
  "[Port]PointerLeave",
  "[Port]PointerEnter",
  "[Port]PointerLeave",
  "[Edge]ConnectEnd",
]
`;

exports[`test Connecting should attach to port: events 1`] = `
Array [
  "[Canvas]ViewportResize",
  "[Canvas]VirtualizationRecalculated",
  "[Canvas]ViewportResize",
  "[Edge]ConnectStart",
  "[Port]PointerEnter",
  "[Port]PointerLeave",
]
`;

exports[`test Connecting should not connect: events 1`] = `
Array [
  "[Canvas]ViewportResize",
  "[Canvas]VirtualizationRecalculated",
  "[Canvas]ViewportResize",
  "[Edge]ConnectStart",
  "[Edge]ConnectEnd",
]
`;

exports[`test Connecting should render invisible lines 1`] = `undefined`;

exports[`test Connecting should render invisible lines: events 1`] = `
Array [
  "[Canvas]ViewportResize",
  "[Canvas]VirtualizationRecalculated",
  "[Canvas]ViewportResize",
]
`;

exports[`test getNearestConnectablePort should return connectable port: events 1`] = `Array []`;

exports[`test getNearestConnectablePort should return nothing with no ports: events 1`] = `Array []`;
